In [1]:
#Commands for downgrading Python version in Kaggle notebook

#adds external APT repository
!add-apt-repository -y 'ppa:deadsnakes/ppa'

#installs specific python version
!apt install -y python3.10

#installs distutils package for the specific python version
!apt-get install -y python3.10-distutils

#Install Python modules using the newly installed Python version
#-I, Ignores and overwrites the installed packages.
!sudo /usr/bin/python3.10 -m pip install -Iv <package-name>

#Run Python scripts using the newly installed Python version
!/usr/bin/python3.10 <python_script.py>

Repository: 'deb https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu/ jammy main'
Description:
This PPA contains more recent Python versions packaged for Ubuntu.

Disclaimer: there's no guarantee of timely updates in case of security problems or other issues. If you want to use them in a security-or-otherwise-critical environment (say, on a production server), you do so at your own risk.

Update Note
Please use this repository instead of ppa:fkrull/deadsnakes.

Reporting Issues

Issues can be reported in the master issue tracker at:
https://github.com/deadsnakes/issues/issues

Supported Ubuntu and Python Versions

- Ubuntu 20.04 (focal) Python3.5 - Python3.7, Python3.9 - Python3.13
- Ubuntu 22.04 (jammy) Python3.7 - Python3.9, Python3.11 - Python3.13
- Ubuntu 24.04 (noble) Python3.7 - Python3.11, Python3.13
- Note: Python2.7 (focal, jammy), Python 3.8 (focal), Python 3.10 (jammy), Python3.12 (noble) are not provided by deadsnakes as upstream ubuntu provides those packages.

Why some

In [2]:
!python --version

Python 3.11.11


In [3]:
!pip install pyspark
!pip install lightning
!pip install pytorch-lifestream

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 819.0/819.0 kB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.7 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.6 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 8.0 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 31.4 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 13.5 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 7.8 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 15.2 MB/s eta 0:00:00:00:0100:01
  Attempting uninstall: packaging
    Found existing installation: packaging 25.0
    Uninstalling packaging-25.0:
      Successfully uninstalled packaging-25.0
  Attempting uninstall: nvidia-nvjitlink

In [4]:
import wandb

wandb.login(key="79f2120f8d4212aceb2c60b3c89a1b6727c19cff")

wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.
wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: toly-kiri4enko (bstu) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

In [5]:
new_config = """#!/usr/bin/env bash

mkdir -p data
cd data/
mkdir -p raw_data
cd raw_data/

curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/0433a4ca/transactions.zip
curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/0554f0cf/clickstream.zip
curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/acfacf11/train_matching.csv

curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/b949c04c/mcc_codes.csv
curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/705abbab/click_categories.csv
curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/e33f2201/currency_rk.csv

curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/a3643657/sample_submission.zip
curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/24687252/baseline_catboost.zip

curl -OL https://storage.yandexcloud.net/datasouls-ods/materials/b99fed70/puzzle.csv

cd ../../
"""

with open('/kaggle/working/get_data.sh', mode='w') as file:
    file.write(new_config)

In [6]:
!source /kaggle/working/get_data.sh

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  250M  100  250M    0     0  50.8M      0  0:00:04  0:00:04 --:--:-- 56.7M
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  469M  100  469M    0     0  48.7M      0  0:00:09  0:00:09 --:--:-- 51.6M
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1045k  100 1045k    0     0  2062k      0 --:--:-- --:--:-- --:--:-- 2061k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  195k  100  195k    0     0   538k      0 --:--:-- --:--:-- --:--:--  540k
  % Total    % Received % Xferd  Average Speed   Tim

# Prepare Data

In [7]:
!mkdir data/train_matching_folds
!mkdir data/transactions_folds
!mkdir data/clickstream_folds

In [8]:
import os

import pyspark
from pyspark.sql import SparkSession
import pyspark.sql.functions as F
from pyspark.sql import types as T
import ptls
from ptls.preprocessing import PysparkDataPreprocessor

# os.environ['JAVA_HOME']= ''

spark_conf = pyspark.SparkConf()
spark_conf.setMaster("local[*]").setAppName("JoinModality")
spark_conf.set("spark.driver.maxResultSize", "8g")
spark_conf.set("spark.executor.memory", "16g")
spark_conf.set("spark.executor.memoryOverhead", "8g")
spark_conf.set("spark.driver.memory", "16g")
spark_conf.set("spark.driver.memoryOverhead", "8g")
spark_conf.set("spark.cores.max", "4")
spark_conf.set("spark.sql.shuffle.partitions", "200")
spark_conf.set("spark.local.dir", "../../spark_local_dir")


spark = SparkSession.builder.config(conf=spark_conf).getOrCreate()
spark.sparkContext.getConf().getAll()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/05/25 00:24:00 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
25/05/25 00:24:00 WARN SparkConf: Note that spark.local.dir will be overridden by the value set by the cluster manager (via SPARK_LOCAL_DIRS in mesos/standalone/kubernetes and LOCAL_DIRS in YARN).


[('spark.driver.extraJavaOptions',
  '-Djava.net.preferIPv6Addresses=false -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/jdk.internal.ref=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Djdk.reflect.useDirectMethodHandle=false'),
 ('spark.app.name', 'JoinModality'),
 ('spark.local.dir', '../../spark_local_dir'),
 ('spark.executor.i

In [ ]:
!unzip ./data/raw_data/transactions.zip -d ./data/raw_data/
!unzip ./data/raw_data/clickstream.zip -d ./data/raw_data/

Archive:  ./data/raw_data/transactions.zip
  inflating: ./data/raw_data/transactions.csv  
  inflating: ./data/raw_data/__MACOSX/._transactions.csv  
Archive:  ./data/raw_data/clickstream.zip
  inflating: ./data/raw_data/clickstream.csv  

In [ ]:
transactions = spark.read.csv('./data/raw_data/transactions.csv', header=True)
clickstream = spark.read.csv('./data/raw_data/clickstream.csv', header=True)
train_matching = spark.read.csv('./data/raw_data/train_matching.csv', header=True)
# train_edu = spark.read.csv('../data/raw_data/train_edu.csv', header=True) #Этого файла нет

click_categories = spark.read.csv('./data/raw_data/click_categories.csv', header=True)
clickstream = clickstream.join(click_categories, on='cat_id')

# Prepare TRX

In [ ]:
preprocessor_trx = PysparkDataPreprocessor(
        col_id='user_id',
        col_event_time='transaction_dttm',
        event_time_transformation='dt_to_timestamp',
        cols_category=["mcc_code", "currency_rk"],
    )

In [ ]:
transactions_prepared = preprocessor_trx.fit_transform(transactions)

In [ ]:
# train_matching = train_matching.withColumnRenamed('rtk', 'user_id')
# clickstream_prepared = clickstream_prepared.join(train_matching, on='user_id', how='outer').drop('user_id')
# clickstream_prepared  = clickstream_prepared.withColumnRenamed('bank', 'user_id')
# clickstream_prepared.show(2)

In [ ]:
# clickstream_prepared = clickstream_prepared.withColumnRenamed('event_time', 'click_event_time')
# transactions_prepared = transactions_prepared.withColumnRenamed('event_time', 'trx_event_time')

In [ ]:
# mm_dataset = transactions_prepared.join(clickstream_prepared, on='user_id', how='outer')
# mm_dataset.show(5)

In [ ]:
!mkdir spark_data

In [ ]:
# mm_dataset.write.mode('overwrite').parquet('./spark_data/mm_dataset.parquet')

In [ ]:
transactions_fold_0, transactions_fold_1, transactions_fold_2,  transactions_fold_3, transactions_fold_4 = transactions_prepared.randomSplit([0.2, 0.2, 0.2, 0.2, 0.2], seed=42)
transactions_fold_0.write.mode('overwrite').parquet('./spark_data/trx_dataset_fold/fold=0')
transactions_fold_1.write.mode('overwrite').parquet('./spark_data/trx_dataset_fold/fold=1')
transactions_fold_2.write.mode('overwrite').parquet('./spark_data/mm_dataset_fold/fold=2')
transactions_fold_3.write.mode('overwrite').parquet('./spark_data/trx_dataset_fold/fold=3')
transactions_fold_4.write.mode('overwrite').parquet('./spark_data/trx_dataset_fold/fold=4')

In [ ]:
del transactions_fold_0
del transactions_fold_1
del transactions_fold_2
del transactions_fold_3
del transactions_fold_4

# Prepare Clickstream

In [ ]:
preprocessor_click = PysparkDataPreprocessor(
    col_id='user_id',
    col_event_time='timestamp',
    event_time_transformation='dt_to_timestamp',
    cols_category=['cat_id', 'level_0', 'level_1', 'level_2'],
)
clickstream_prepared = preprocessor_click.fit_transform(clickstream)

In [ ]:
clickstream_fold_0, clickstream_fold_1, clickstream_fold_2, clickstream_fold_3, clickstream_fold_4 = clickstream_prepared.randomSplit([0.2, 0.2, 0.2, 0.2, 0.2], seed=42)
clickstream_fold_0.write.mode('overwrite').parquet('./spark_data/click_dataset_fold/fold=0')
clickstream_fold_1.write.mode('overwrite').parquet('./spark_data/click_dataset_fold/fold=1')
clickstream_fold_2.write.mode('overwrite').parquet('./spark_data/mm_dataset_fold/fold=2')
clickstream_fold_3.write.mode('overwrite').parquet('./spark_data/click_dataset_fold/fold=3')
clickstream_fold_4.write.mode('overwrite').parquet('./spark_data/click_dataset_fold/fold=4')

In [ ]:
# datetimes = []
# for idx, item in enumerate(iter(trx_train)):
#     if idx == 100:
#         break
#     datetimes.extend(pd.to_datetime(item['event_time'].numpy(), unit='s'))

# for target_month in range(1, 13):
#     filtered_dates = [dt for dt in datetimes if dt.month == target_month]
    
#     year = 2021
#     _, num_days = calendar.monthrange(year, target_month)
    
#     days_in_month = [dt.day for dt in filtered_dates]
#     day_counts = pd.Series(days_in_month).value_counts().sort_index()
    
#     day_counts = day_counts.reindex(range(1, num_days + 1), fill_value=0)
    
#     plt.figure(figsize=(14, 6))
#     plt.bar(
#         day_counts.index.astype(str),
#         day_counts.values,
#         color='skyblue',
#         edgecolor='black'
#     )
    
#     plt.title(f'Распределение меток по дням месяца ({datetime(year, target_month, 1).strftime("%B %Y")})', fontsize=14)
#     plt.xlabel('День месяца', fontsize=12)
#     plt.ylabel('Количество меток', fontsize=12)
#     plt.xticks(rotation=45)
#     plt.grid(axis='y', linestyle='--', alpha=0.7)
#     plt.show()

In [ ]:
# import matplotlib.pyplot as plt
# import pandas as pd
# import torch
# from datetime import datetime
# import calendar

# for idx, item in enumerate(iter(trx_train)):
#     if idx == 30:
#         break

#     datetimes = pd.to_datetime(item['event_time'].numpy(), unit='s')
    
#     target_month = 2 
#     filtered_dates = [dt for dt in datetimes if dt.month == target_month]

#     year = 2021
#     _, num_days = calendar.monthrange(year, target_month)

#     days_in_month = [dt.day for dt in filtered_dates]
#     day_counts = pd.Series(days_in_month).value_counts().sort_index()

#     day_counts = day_counts.reindex(range(1, num_days + 1), fill_value=0)

#     plt.figure(figsize=(14, 6))
#     plt.bar(
#         day_counts.index.astype(str),
#         day_counts.values,
#         color='skyblue',
#         edgecolor='black'
#     )
    
#     plt.title(f'Распределение меток по дням месяца ({datetime(year, target_month, 1).strftime("%B %Y")})', fontsize=14)
#     plt.xlabel('День месяца', fontsize=12)
#     plt.ylabel('Количество меток', fontsize=12)
#     plt.xticks(rotation=45)
#     plt.grid(axis='y', linestyle='--', alpha=0.7)
#     plt.show()

In [ ]:
del clickstream_fold_0
del clickstream_fold_1
del clickstream_fold_2
del clickstream_fold_3
del clickstream_fold_4

In [ ]:
del transactions_prepared
del clickstream_prepared

In [ ]:
spark.stop()

# Utilities

In [ ]:
from ptls.data_load.iterable_processing_dataset import IterableProcessingDataset
from datetime import datetime
from ptls.data_load.padded_batch import PaddedBatch

class DeleteNan(IterableProcessingDataset):
    def __init__(self, col_name):
        super().__init__()
        self.col_name = col_name
    
    def __iter__(self):
        for rec in self._src:
            features = rec[0] if type(rec) is tuple else rec
            if features[self.col_name] is not None:
                yield features

class TypeProc(IterableProcessingDataset):
    def __init__(self, col_name, tp='float'):
        super().__init__()
        self.col_name = col_name
        self.tp = tp

    def __iter__(self):
        for rec in self._src:
            features = rec[0] if type(rec) is tuple else rec
            if type(features[self.col_name]) is not str:
                features[self.col_name] = np.array([float(val) if self.tp=='float' else int(float(val)) for val in features[self.col_name]])
            else:
                features[self.col_name] = float(features[self.col_name]) if self.tp=='float' else int(float(features[self.col_name]))
            yield features

# TRX Baseline

In [ ]:
from ptls.data_load.datasets import ParquetDataset
from ptls.data_load.iterable_processing.time_proc import TimeProc
import numpy as np

trx_train = ParquetDataset(
    shuffle_files=True,
    data_files=[
        './spark_data/trx_dataset_fold/fold=0',
        './spark_data/trx_dataset_fold/fold=1',
        # './spark_data/trx_dataset_fold/fold=2',
        './spark_data/trx_dataset_fold/fold=3'
    ],
    i_filters=[
        DeleteNan(col_name='mcc_code'),
        TypeProc(col_name='transaction_amt'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            drop_feature_names=[
                'user_id',
            ]
        ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'mcc_code': 349,
                'currency_rk': 9
            }
        ),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

trx_valid = ParquetDataset(
    shuffle_files=False,
    data_files=[
        './spark_data/trx_dataset_fold/fold=4'
    ],
    i_filters=[
        DeleteNan(col_name='mcc_code'),
        TypeProc(col_name='transaction_amt'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            drop_feature_names=[
                'user_id',
            ]
        ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'mcc_code': 349,
                'currency_rk': 9
            }
        ),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

In [29]:
# dates = []

# for idx, item in enumerate(iter(trx_train)):
#     if idx == 100:
#         break

#     dates.extend(pd.to_datetime(item['event_time'].numpy(), unit='s'))

# dates = pd.DatetimeIndex(dates)

# days_of_week = dates.dayofweek.tolist()

# days_counts = pd.Series(days_of_week).value_counts().sort_index()

# days_counts = days_counts.reindex(range(7), fill_value=0)

# weekday_names = [
#     'Понедельник', 'Вторник', 'Среда',
#     'Четверг', 'Пятница', 'Суббота', 'Воскресенье'
# ]

# plt.figure(figsize=(10, 6))
# plt.bar(weekday_names, days_counts, color='skyblue')
# plt.title('Распределение меток по дням недели', fontsize=14)
# plt.xlabel('День недели', fontsize=12)
# plt.ylabel('Количество меток', fontsize=12)
# plt.xticks(rotation=45)
# plt.grid(axis='y', linestyle='--', alpha=0.7)
# plt.show()

In [30]:
# import matplotlib.pyplot as plt
# import pandas as pd

# for idx, item in enumerate(iter(trx_train)):
#     if idx == 30:
#         break
#     dates = pd.to_datetime(item['event_time'].numpy(), unit='s')
    
#     days_of_week = dates.dayofweek.tolist()
    
#     days_counts = pd.Series(days_of_week).value_counts().sort_index()
    
#     days_counts = days_counts.reindex(range(7), fill_value=0)
    
#     weekday_names = [
#         'Понедельник', 'Вторник', 'Среда',
#         'Четверг', 'Пятница', 'Суббота', 'Воскресенье'
#     ]
    
#     plt.figure(figsize=(10, 6))
#     plt.bar(weekday_names, days_counts, color='skyblue')
#     plt.title('Распределение меток по дням недели', fontsize=14)
#     plt.xlabel('День недели', fontsize=12)
#     plt.ylabel('Количество меток', fontsize=12)
#     plt.xticks(rotation=45)
#     plt.grid(axis='y', linestyle='--', alpha=0.7)
#     plt.show()

In [31]:
from ptls.frames.coles import ColesIterableDataset

trx_data_module = ptls.frames.PtlsDataModule(
    train_data=ColesIterableDataset(
        data=trx_train,
        col_time='event_time',
        splitter=ptls.frames.coles.split_strategy.SampleSlices(
            split_count=5,
            cnt_min=32,
            cnt_max=180
        )
    ),
    valid_data=ColesIterableDataset(
        data=trx_valid,
        col_time='event_time',
        splitter=ptls.frames.coles.split_strategy.SampleSlices(
            split_count=5,
            cnt_min=32,
            cnt_max=180
        )
    ),
    train_batch_size=512,
    train_num_workers=2,
    valid_batch_size=128,
    valid_num_workers=2
)

2025-05-25 00:05:45.788210: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748131546.031940      35 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748131546.106713      35 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [32]:
from ptls.frames.coles.metric import BatchRecallTopK
from functools import partial
import torch

trx_encoder = ptls.nn.TrxEncoder(
    norm_embeddings=False,
    embeddings_noise=0.003,
    embeddings={
        'mcc_code': {'in': 350, 'out': 64},
        'currency_rk': {'in': 10, 'out': 4},
        'hour': {'in': 25, 'out': 16},
        'weekday': {'in': 8, 'out': 4}
    },
    numeric_values={
        'transaction_amt': 'log'
    }
)

seq_encoder = ptls.nn.RnnSeqEncoder(
    trx_encoder=trx_encoder,
    type='gru',
    hidden_size=256
)

optimizer_partial = partial(torch.optim.AdamW, lr=0.001, weight_decay=1e-4)
lr_scheduler_partial = partial(torch.optim.lr_scheduler.StepLR, step_size=3, gamma=0.9025)

trx_pl_module = ptls.frames.coles.CoLESModule(
    validation_metric=BatchRecallTopK(
        K=4,
        metric='cosine'
    ),
    seq_encoder=seq_encoder,
    optimizer_partial=optimizer_partial,
    lr_scheduler_partial=lr_scheduler_partial
)

In [33]:
from lightning.pytorch.loggers import WandbLogger
import pytorch_lightning as pl
import numpy as np

wandb_logger = WandbLogger(project="MBD_My_Code", log_model="all", name="df2022_trx_base")

trainer = pl.Trainer(
    logger=wandb_logger,
    max_epochs=30,
    accelerator="cuda" if torch.cuda.is_available() else "cpu",
    enable_progress_bar=True,
    gradient_clip_val=0.5,
    log_every_n_steps=50,
    limit_val_batches=32
)

INFO: Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
INFO: GPU available: True (cuda), used: True
INFO: TPU available: False, using: 0 TPU cores
INFO: HPU available: False, using: 0 HPUs


In [34]:
trainer.fit(trx_pl_module, trx_data_module)

Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO: `Trainer.fit` stopped: `max_epochs=30` reached.


# Click baseline

In [56]:
from ptls.data_load.datasets import ParquetDataset
from ptls.data_load.iterable_processing.time_proc import TimeProc
import numpy as np

click_train = ParquetDataset(
    shuffle_files=True,
    data_files=[
        './spark_data/click_dataset_fold/fold=0',
        './spark_data/click_dataset_fold/fold=1',
        # './spark_data/click_dataset_fold/fold=2',
        './spark_data/click_dataset_fold/fold=3'
    ],
    i_filters=[
        DeleteNan(col_name='cat_id'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            drop_feature_names=[
                'user_id',
                'new_uid'
            ]
        ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'cat_id': 399,
                'level_0': 399,
                'level_1': 399,
                'hour': 24,
                'weekday': 7
            }
        ),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

click_valid = ParquetDataset(
    shuffle_files=True,
    data_files=[
        './spark_data/click_dataset_fold/fold=4'
    ],
    i_filters=[
        DeleteNan(col_name='cat_id'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            drop_feature_names=[
                'user_id',
                'new_uid'
            ]
        ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'cat_id': 399,
                'level_0': 399,
                'level_1': 399,
                'hour': 24,
                'weekday': 7
            }
        ),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

In [57]:
next(iter(click_train))

{'event_time': tensor([1625326320, 1625326380, 1625326800,  ..., 1628749983, 1628779831,
         1628947292]),
 'cat_id': tensor([6, 6, 6,  ..., 7, 7, 6], dtype=torch.int32),
 'level_0': tensor([7, 7, 7,  ..., 8, 8, 7], dtype=torch.int32),
 'level_1': tensor([1, 1, 1,  ..., 1, 1, 1], dtype=torch.int32),
 'level_2': tensor([1, 1, 1,  ..., 1, 1, 1], dtype=torch.int32),
 'weekday': tensor([6, 6, 6,  ..., 4, 4, 6]),
 'hour': tensor([15, 15, 15,  ...,  6, 14, 13])}

In [62]:
from ptls.frames.coles import ColesIterableDataset

click_data_module = ptls.frames.PtlsDataModule(
    train_data=ColesIterableDataset(
        data=click_train,
        col_time='event_time',
        splitter=ptls.frames.coles.split_strategy.SampleSlices(
            split_count=5,
            cnt_min=32,
            cnt_max=180
        )
    ),
    valid_data=ColesIterableDataset(
        data=click_valid,
        col_time='event_time',
        splitter=ptls.frames.coles.split_strategy.SampleSlices(
            split_count=5,
            cnt_min=32,
            cnt_max=180
        )
    ),
    train_batch_size=512,
    train_num_workers=2,
    valid_batch_size=128,
    valid_num_workers=2
)

In [63]:
from ptls.frames.coles.metric import BatchRecallTopK
from functools import partial
import torch

click_encoder = ptls.nn.TrxEncoder(
    norm_embeddings=False,
    embeddings_noise=0.003,
    embeddings={
        'cat_id': {'in': 400, 'out': 64},
        'level_0': {'in': 400, 'out': 16},
        'level_1': {'in': 400, 'out': 8},
        'level_2': {'in': 400, 'out': 4},
        'hour': {'in': 25, 'out': 16},
        'weekday': {'in': 8, 'out': 4}
    }
)

seq_encoder = ptls.nn.RnnSeqEncoder(
    trx_encoder=click_encoder,
    type='gru',
    hidden_size=256
)

optimizer_partial = partial(torch.optim.AdamW, lr=0.001, weight_decay=1e-4)
lr_scheduler_partial = partial(torch.optim.lr_scheduler.StepLR, step_size=3, gamma=0.9025)

click_pl_module = ptls.frames.coles.CoLESModule(
    validation_metric=BatchRecallTopK(
        K=4,
        metric='cosine'
    ),
    seq_encoder=seq_encoder,
    optimizer_partial=optimizer_partial,
    lr_scheduler_partial=lr_scheduler_partial
)

In [64]:
from lightning.pytorch.loggers import WandbLogger
import pytorch_lightning as pl
import numpy as np

wandb_logger = WandbLogger(project="MBD_My_Code", log_model="all")

trainer = pl.Trainer(
    logger=wandb_logger,
    max_epochs=30,
    accelerator="cuda" if torch.cuda.is_available() else "cpu",
    enable_progress_bar=True,
    gradient_clip_val=0.5,
    log_every_n_steps=50,
    limit_val_batches=32
)

INFO: You are using the plain ModelCheckpoint callback. Consider using LitModelCheckpoint which with seamless uploading to Model registry.
INFO: GPU available: True (cuda), used: True
INFO: TPU available: False, using: 0 TPU cores
INFO: HPU available: False, using: 0 HPUs


In [65]:
trainer.fit(click_pl_module, click_data_module)

Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7eb911af45e0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1604, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1587, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.11/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7eb911af45e0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 1604, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py", line 15

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO: `Trainer.fit` stopped: `max_epochs=30` reached.


In [69]:
del click_train
del click_valid
del trx_train
del trx_valid

In [76]:
import gc

gc.collect()

8

**inference**

In [74]:
dataset_trx = ptls.data_load.datasets.ParquetDataset(
    data_files=['./spark_data/trx_dataset_fold/fold=0',
               './spark_data/trx_dataset_fold/fold=1',
               './spark_data/trx_dataset_fold/fold=3',
               './spark_data/trx_dataset_fold/fold=4'
               ],
    i_filters=[
        DeleteNan(col_name='mcc_code'),
        DeleteNan(col_name='user_id'),
        TypeProc(col_name='transaction_amt'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            keep_feature_names=[
                'user_id'
            ]
        ),
        ptls.data_load.iterable_processing.iterable_seq_len_limit.ISeqLenLimit(max_seq_len=4096),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'mcc_code': 349,
                'currency_rk': 9
            }
        ),
        ptls.data_load.iterable_processing.time_proc.TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.to_torch_tensor.ToTorch()
    ]
)

dataset_click = ptls.data_load.datasets.ParquetDataset(
    data_files=[
        './spark_data/click_dataset_fold/fold=0',
        './spark_data/click_dataset_fold/fold=1',
        # './spark_data/click_dataset_fold/fold=2',
        './spark_data/click_dataset_fold/fold=3'
    ],
    i_filters=[
        DeleteNan(col_name='cat_id'),
        DeleteNan(col_name='user_id'),
        ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
            drop_feature_names=[
                'new_uid'
            ],
            keep_feature_names=[
                'user_id'
            ]
        ),
        ptls.data_load.iterable_processing.iterable_seq_len_limit.ISeqLenLimit(max_seq_len=4096),
        ptls.data_load.iterable_processing.time_proc.TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'cat_id': 399,
                'level_0': 399,
                'level_1': 399,
                'hour': 24,
                'weekday': 7
            }
        ),
        ptls.data_load.iterable_processing.to_torch_tensor.ToTorch()
    ]
)

In [77]:
import pandas as pd
import pytorch_lightning as pl
import torch
import numpy as np
from itertools import chain
from ptls.data_load.padded_batch import PaddedBatch
from datetime import datetime
from ptls.custom_layers import StatPooling
from ptls.nn.seq_step import LastStepEncoder


class InferenceModuleMultimodal(pl.LightningModule):
    def __init__(
        self,
        model,
        pandas_output=True,
        col_id='user_id',
        target_col_names=None,
        model_out_name='emb',
        model_type='notab'
    ):
        super().__init__()

        self.model = model
        self.pandas_output = pandas_output
        self.target_col_names = target_col_names
        self.col_id = col_id
        self.model_out_name = model_out_name
        self.model_type = model_type

        self.stat_pooler = StatPooling()
        self.last_step = LastStepEncoder()

    def forward(self, x):
        x_len = len(x)
        if x_len == 3:
            x, batch_ids, target_cols = x
        else: 
            x, batch_ids = x
        if 'seq_encoder' in dir(self.model):
            out = self.model.seq_encoder(x)
        else:
            out = self.model(x)
            
        if x_len == 3:
            target_cols = torch.tensor(target_cols)
            x_out = {
                self.col_id: batch_ids,
                self.model_out_name: out
            }
            if len(target_cols.size()) > 1:
                for idx, target_col in enumerate(self.target_col_names):
                    x_out[target_col] = target_cols[:, idx]
            else: 
                x_out[self.target_col_names[0]] = target_cols[::4]
        else:
            x_out = {
                self.col_id: batch_ids,
                self.model_out_name: out
            }

        if self.pandas_output:
            return self.to_pandas(x_out)
        return x_out

    @staticmethod
    def to_pandas(x):
        expand_cols = []
        scalar_features = {}

        for k, v in x.items():
            if type(v) is torch.Tensor:
                v = v.cpu().numpy()

            if type(v) is list or len(v.shape) == 1:
                scalar_features[k] = v
            elif len(v.shape) == 2:
                expand_cols.append(k)
            else:
                scalar_features[k] = None

        dataframes = [pd.DataFrame(scalar_features)]
        for col in expand_cols:
            v = x[col].cpu().numpy()
            dataframes.append(pd.DataFrame(v, columns=[f'{col}_{i:04d}' for i in range(v.shape[1])]))

        return pd.concat(dataframes, axis=1)

def collate_feature_dict(batch):
    new_x_ = defaultdict(list)
    for i, x in enumerate(batch):
        for k, v in x.items():
            new_x_[k].append(v)
    
    seq_col = next(k for k, v in batch[0].items() if FeatureDict.is_seq_feature(k, v))
    lengths = torch.LongTensor([len(rec[seq_col]) for rec in batch])
    new_x = {}
    for k, v in new_x_.items():
        if type(v[0]) is torch.Tensor:
            if k.startswith('target'):
                new_x[k] = torch.stack(v, dim=0)
            else:
                new_x[k] = torch.nn.utils.rnn.pad_sequence(v, batch_first=True)
        elif type(v[0]) is np.ndarray:
            new_x[k] = v  # list of arrays[object]
        else:
            v = np.array(v)
            if v.dtype.kind == 'i':
                new_x[k] = torch.from_numpy(v).long()
            elif v.dtype.kind == 'f':
                new_x[k] = torch.from_numpy(v).float()
            elif v.dtype.kind == 'b':
                new_x[k] = torch.from_numpy(v).bool()
            else:
                new_x[k] = v
    return PaddedBatch(new_x, lengths)


def collate_feature_dict_with_target(batch, col_id='user_id', target_col_names=None):
    batch_ids = []
    target_cols = []
    for sample in batch:
        batch_ids.append(sample[col_id])
        del sample[col_id]
        
        if target_col_names is not None:
            sample_targets = []
            for target_col in target_col_names:
                sample_targets.append(sample[target_col])
                del sample[target_col]
            target_cols.append(sample_targets)
                
            
    padded_batch = collate_feature_dict(batch)
    if target_col_names is not None:
        return padded_batch, batch_ids, target_cols
    return padded_batch, batch_ids

In [78]:
from torch.utils.data import Dataset, DataLoader

target_col_names = [
    'target_1',
    'target_2',
    'target_3',
    'target_4'
]

collate_fn = partial(
    collate_feature_dict_with_target,
    target_col_names=None
)

trx_dl = DataLoader(
    dataset=dataset_trx,
    collate_fn=collate_fn,
    shuffle=False,
    num_workers=0,
    batch_size=32
)
click_dl = DataLoader(
    dataset=dataset_click,
    collate_fn=collate_fn,
    shuffle=False,
    num_workers=0,
    batch_size=32
)

In [79]:
trx_inf_module = InferenceModuleMultimodal(
    model=trx_pl_module,
    pandas_output=True,
    col_id='user_id',
    target_col_names=None
)

click_inf_module = InferenceModuleMultimodal(
    model=click_pl_module,
    pandas_output=True,
    col_id='user_id',
    target_col_names=None
)

In [181]:
from collections import defaultdict
from ptls.data_load.feature_dict import FeatureDict

trx_embeddings = pd.concat(trainer.predict(trx_inf_module, trx_dl))
click_embeddings = pd.concat(trainer.predict(click_inf_module, click_dl))

/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: The 'predict_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.


Predicting: |          | 0/? [00:00<?, ?it/s]

/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: The 'predict_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.


Predicting: |          | 0/? [00:00<?, ?it/s]

In [182]:
train_matching = pd.read_csv('./data/raw_data/train_matching.csv')

In [183]:
train_matching

bank                               rtk
0      178b387813ac4a63a2200274a9e3876e  e19c0f141e9442c5bdb0d5751f6fba1c
1      47cffa46e6b04389ba91a70735d2c3ff  7df3371aabd349e4ac2de8299238627d
2      f73b767cfd72472aa06df9a69e746aaf  b23d62c7e41145a7ad140a4fcfaafefa
3      48da649603734185b98dcea93484dbc7  63ad789541c54463ad54d0a707b3ab8b
4      37304ef19de542ee8fe8a6f050236525  c0e96de5dd594d948a8b9fde9e6980cb
...                                 ...                               ...
17576  d870a0dfad91413cae05f001c14dd212                                 0
17577  be367119b35e4dbc8e6cb07221822f57                                 0
17578  518dccd73b844e18b67a71f2dae925cd                                 0
17579  43d30f9a20264a79b3e212c1a452be95                                 0
17580  54310a61cfb64c4aaa0003ecdb727bf8                                 0

[17581 rows x 2 columns]

In [184]:
def gen_random_for_negative(x, k, list_of_uniq):
    while True:
        final_list = np.random.choice(list_of_uniq, size=k, replace=False)
        if x not in final_list:
            return final_list

In [185]:
list_of_uniq_rtk = train_matching['rtk'].unique()
list_of_uniq_bank = train_matching['bank'].unique()
train_matching['negatives'] = train_matching['rtk'].apply(
                        lambda x: gen_random_for_negative(x, 3, list_of_uniq_rtk))

In [186]:
train_matching

bank                               rtk  \
0      178b387813ac4a63a2200274a9e3876e  e19c0f141e9442c5bdb0d5751f6fba1c   
1      47cffa46e6b04389ba91a70735d2c3ff  7df3371aabd349e4ac2de8299238627d   
2      f73b767cfd72472aa06df9a69e746aaf  b23d62c7e41145a7ad140a4fcfaafefa   
3      48da649603734185b98dcea93484dbc7  63ad789541c54463ad54d0a707b3ab8b   
4      37304ef19de542ee8fe8a6f050236525  c0e96de5dd594d948a8b9fde9e6980cb   
...                                 ...                               ...   
17576  d870a0dfad91413cae05f001c14dd212                                 0   
17577  be367119b35e4dbc8e6cb07221822f57                                 0   
17578  518dccd73b844e18b67a71f2dae925cd                                 0   
17579  43d30f9a20264a79b3e212c1a452be95                                 0   
17580  54310a61cfb64c4aaa0003ecdb727bf8                                 0   

                                               negatives  
0      [06cedac1877d42f3b9fda560130ff642, 28defcd067e...  
1      [87aa30cde2f14aa39e4e2f111e443feb, ebc1b4ee96a...  
2      [750f4542177244cc9673bd6e37a57768, 9bd736e62a2...  
3      [0b76cdb34090489282c7d53489c17b98, ab6060fa24f...  
4      [705d843fe89a4599a27fb44137f50e64, 2ae63656751...  
...                                                  ...  
17576  [a113e5041d9947bda1533ef6fbf44afd, 9d53538afc9...  
17577  [1d9aa2bb649a4849b7a1306a1b017e6b, 0119028e776...  
17578  [b29184f9e1e349f9990d64fc6668f0dc, eeb8f921e85...  
17579  [bcab38c5a7cb407e8891baea88b28a01, def8c449a33...  
17580  [e2a02bc05fae4118b3c4284a3b4d9ffa, b1a6c710333...  

[17581 rows x 3 columns]

In [187]:
train_matching['target'] = 1
pos_part = train_matching[['bank', 'rtk', 'target']]
neg_part = train_matching[['bank', 'negatives']].explode('negatives')
neg_part['target'] = 0
neg_part.columns = pos_part.columns
full_dataset = pd.concat([pos_part, neg_part]).sort_values(by='bank')

In [188]:
full_dataset

bank                               rtk  \
3418   000932580e404dafbecd5916d4640938  76f99811d8314f4380987edae155e990   
3418   000932580e404dafbecd5916d4640938  bcca2e8a490a413aa7e1dbd154f04abf   
3418   000932580e404dafbecd5916d4640938  b11c550b7bbd465d8642ae53ae71f758   
3418   000932580e404dafbecd5916d4640938  97d2092878ea426788070edd04107a9d   
5891   0009e36b42cb4caeb928a1e596819495  beed41e945754ac5ad6af2014d759c3a   
...                                 ...                               ...   
3293   fff9967567c14326bb85d29009e86aea  c8a9623e62664799816b32ce50e46c3e   
13090  fffb051877b949f49bed6538393a06ae  7fa88b9066fc41878c28cc192b0005cb   
13090  fffb051877b949f49bed6538393a06ae  565727ce66ea4ef48b6ad40b3d468f94   
13090  fffb051877b949f49bed6538393a06ae  fdd45c71664547fe9b7fd632720ceb66   
13090  fffb051877b949f49bed6538393a06ae  1466f7dcb1c4400aaab226287e17e7c8   

       target  
3418        0  
3418        0  
3418        0  
3418        1  
5891        1  
...       ...  
3293        0  
13090       0  
13090       0  
13090       0  
13090       1  

[70324 rows x 3 columns]

In [189]:
trx_embeddings = trx_embeddings.rename(columns={'user_id': 'bank'})

In [190]:
click_embeddings = click_embeddings.rename(columns={'user_id': 'rtk'})

In [191]:
full_dataset

bank                               rtk  \
3418   000932580e404dafbecd5916d4640938  76f99811d8314f4380987edae155e990   
3418   000932580e404dafbecd5916d4640938  bcca2e8a490a413aa7e1dbd154f04abf   
3418   000932580e404dafbecd5916d4640938  b11c550b7bbd465d8642ae53ae71f758   
3418   000932580e404dafbecd5916d4640938  97d2092878ea426788070edd04107a9d   
5891   0009e36b42cb4caeb928a1e596819495  beed41e945754ac5ad6af2014d759c3a   
...                                 ...                               ...   
3293   fff9967567c14326bb85d29009e86aea  c8a9623e62664799816b32ce50e46c3e   
13090  fffb051877b949f49bed6538393a06ae  7fa88b9066fc41878c28cc192b0005cb   
13090  fffb051877b949f49bed6538393a06ae  565727ce66ea4ef48b6ad40b3d468f94   
13090  fffb051877b949f49bed6538393a06ae  fdd45c71664547fe9b7fd632720ceb66   
13090  fffb051877b949f49bed6538393a06ae  1466f7dcb1c4400aaab226287e17e7c8   

       target  
3418        0  
3418        0  
3418        0  
3418        1  
5891        1  
...       ...  
3293        0  
13090       0  
13090       0  
13090       0  
13090       1  

[70324 rows x 3 columns]

In [192]:
trx_embeddings

bank  emb_0000  emb_0001  emb_0002  emb_0003  \
0   00ba088e17c542bd8039e575ca155048 -0.625001 -0.243617  0.163364  0.480665   
1   0184194d5e104a97868713a7d22aed1f -0.323013 -0.235419 -0.530561  0.152303   
2   01904d1a09ee40468ec3e58dc94b9416 -0.412768 -0.233752 -0.467492  0.082381   
3   020d806d3be9442289bd9a672bf212da  0.000271 -0.277299  0.265619  0.504253   
4   0317073e51314bfd8d987fb98b5277d3 -0.405850 -0.292430  0.109852  0.081532   
..                               ...       ...       ...       ...       ...   
28  f3e653050b484501bee0b32d05778426 -0.650640 -0.192495 -0.683442  0.083355   
29  f4009f4fe6534a308a6b1fd166ad5fc7 -0.314204 -0.267557 -0.495368  0.341328   
30  f7d76e6cd75745f3be5df8290d49a128 -0.249105 -0.262483  0.062181  0.292575   
31  fd74b3d5fbef4297896f2c1892e56685 -0.428595 -0.372882 -0.208509  0.103472   
0   fe528f807b17454bb7a0e721f895398a -0.556809 -0.100131 -0.540976 -0.144857   

    emb_0004  emb_0005  emb_0006  emb_0007  emb_0008  ...  emb_0246  emb_0247  \
0   0.501686  0.101990 -0.044159 -0.313844  0.606088  ...  0.486627 -0.746782   
1   0.714127  0.197821 -0.264800  0.327070  0.505027  ...  0.577334 -0.872298   
2   0.763507 -0.445924 -0.303396  0.435161  0.707241  ...  0.391680 -0.964027   
3   0.481210 -0.064676  0.364488  0.203106  0.589540  ...  0.576690 -0.914879   
4   0.714127 -0.115938  0.094222  0.207996  0.654748  ...  0.562460 -0.808992   
..       ...       ...       ...       ...       ...  ...       ...       ...   
28  0.931046  0.109286  0.175258  0.385544  0.796173  ...  0.259659 -0.599181   
29  0.770136 -0.070886  0.073737  0.004501  0.806574  ...  0.631992 -0.722782   
30  0.761156  0.202873  0.256251  0.247882  0.774876  ... -0.211712 -0.698261   
31  0.463316  0.109624 -0.328438  0.122471  0.714286  ...  0.581544 -0.777710   
0   0.591071  0.243480  0.161219  0.334196  0.681295  ...  0.393105 -0.654121   

    emb_0248  emb_0249  emb_0250  emb_0251  emb_0252  emb_0253  emb_0254  \
0   0.765194 -0.129537  0.412040  0.591943 -0.692759 -0.631624 -0.748908   
1   0.616923 -0.460647  0.773803  0.702861 -0.807951 -0.786805 -0.853459   
2   0.595035  0.298414 -0.149746  0.591051 -0.982628 -0.669345 -0.968200   
3   0.589735 -0.196531  0.134027  0.630934 -0.824845 -0.732586 -0.848887   
4  -0.226964  0.730764  0.510100  0.214419 -0.785357 -0.469446 -0.790325   
..       ...       ...       ...       ...       ...       ...       ...   
28  0.448399  0.692999  0.512021  0.496118 -0.821950 -0.291300 -0.924125   
29  0.688978  0.511670  0.054256  0.376250 -0.958245 -0.601977 -0.968614   
30  0.502025  0.068962  0.328304  0.579600 -0.727478 -0.298953 -0.694111   
31 -0.108956 -0.026611  0.719629  0.585140 -0.696808 -0.581415 -0.821006   
0   0.612554 -0.121475  0.455648  0.653345 -0.579348 -0.455370 -0.802590   

    emb_0255  
0   0.502590  
1   0.546677  
2   0.566579  
3   0.658259  
4  -0.115452  
..       ...  
28  0.601489  
29  0.600059  
30  0.508633  
31  0.460298  
0   0.272393  

[17921 rows x 257 columns]

In [193]:
full_dataset['bank']

3418     000932580e404dafbecd5916d4640938
3418     000932580e404dafbecd5916d4640938
3418     000932580e404dafbecd5916d4640938
3418     000932580e404dafbecd5916d4640938
5891     0009e36b42cb4caeb928a1e596819495
                       ...               
3293     fff9967567c14326bb85d29009e86aea
13090    fffb051877b949f49bed6538393a06ae
13090    fffb051877b949f49bed6538393a06ae
13090    fffb051877b949f49bed6538393a06ae
13090    fffb051877b949f49bed6538393a06ae
Name: bank, Length: 70324, dtype: object

In [194]:
click_embeddings['rtk']

0     0253390d0a30445195c2b6ab4718b432
1     05a7e974c4f143db8eea7a5d23585259
2     05fc62574d064801af8bc2e086a45d7a
3     0a252290b4d44b29bf9f8c855731997e
4     0d8486356bb14e509f4d5c4e0c400abd
                    ...               
7     f9280240b71d42d6b98eda853b963d57
8     f92ce44568bc44dba18e0b25bd25b332
9     fb3b4496b88c4efab486f62e4ad7e563
10    fbeee4870ba147f997ff875f0964077c
11    fdbcae92ce0a42b399d8e7efb7bbb356
Name: rtk, Length: 11820, dtype: object

In [195]:
full_dataset['rtk']

3418     76f99811d8314f4380987edae155e990
3418     bcca2e8a490a413aa7e1dbd154f04abf
3418     b11c550b7bbd465d8642ae53ae71f758
3418     97d2092878ea426788070edd04107a9d
5891     beed41e945754ac5ad6af2014d759c3a
                       ...               
3293     c8a9623e62664799816b32ce50e46c3e
13090    7fa88b9066fc41878c28cc192b0005cb
13090    565727ce66ea4ef48b6ad40b3d468f94
13090    fdd45c71664547fe9b7fd632720ceb66
13090    1466f7dcb1c4400aaab226287e17e7c8
Name: rtk, Length: 70324, dtype: object

In [196]:
trx_embeddings[trx_embeddings['bank'] == '00ba088e17c542bd8039e575ca155048']

bank  emb_0000  emb_0001  emb_0002  emb_0003  \
0  00ba088e17c542bd8039e575ca155048 -0.625001 -0.243617  0.163364  0.480665   

   emb_0004  emb_0005  emb_0006  emb_0007  emb_0008  ...  emb_0246  emb_0247  \
0  0.501686   0.10199 -0.044159 -0.313844  0.606088  ...  0.486627 -0.746782   

   emb_0248  emb_0249  emb_0250  emb_0251  emb_0252  emb_0253  emb_0254  \
0  0.765194 -0.129537   0.41204  0.591943 -0.692759 -0.631624 -0.748908   

   emb_0255  
0   0.50259  

[1 rows x 257 columns]

In [197]:
from tqdm import tqdm

full_data = []
targets = full_dataset['target'].to_numpy()

for trx_id, click_id in tqdm(zip(full_dataset['bank'].to_numpy(), full_dataset['rtk'].to_numpy())):
    trx_emb = trx_embeddings[trx_embeddings['bank'] == str(trx_id)].drop(columns=['bank']).to_numpy().flatten()
    click_emb = click_embeddings[click_embeddings['rtk'] == str(click_id)].drop(columns=['rtk']).to_numpy().flatten()
    if len(trx_emb) > 0 and len(click_emb) > 0:
        full_data.append(np.hstack((trx_emb, click_emb)))

70324it [03:48, 307.73it/s]


In [198]:
full_data = np.stack(full_data)

In [199]:

shuffled_ids = np.random.permutation(len(full_data))

In [200]:
train_ids = shuffled_ids[:int(len(shuffled_ids) * 0.8)]
test_ids = shuffled_ids[int(len(shuffled_ids) * 0.8) + 1:]

In [201]:
len(train_ids), len(test_ids)

(25918, 6479)

In [202]:
X_train, y_train = full_data[train_ids], targets[train_ids]
X_test, y_test = full_data[test_ids], targets[test_ids]


In [203]:
from lightgbm import LGBMClassifier

model = LGBMClassifier(
    n_estimators=1000,
    boosting_type='gbdt',
    subsample=0.5,
    subsample_freq=1,
    learning_rate=0.02,
    feature_fraction=0.75,
    max_depth=6,
    lambda_l1=1,
    lambda_l2=1,
    min_data_in_leaf=50,
    random_state=42,
    n_jobs=8,
    verbose=-1
)

In [ ]:
del full_dataset
del click_embeddings
del trx_embeddings

gc.collect()

In [204]:
model.fit(X_train, y_train)

LGBMClassifier(feature_fraction=0.75, lambda_l1=1, lambda_l2=1,
               learning_rate=0.02, max_depth=6, min_data_in_leaf=50,
               n_estimators=1000, n_jobs=8, random_state=42, subsample=0.5,
               subsample_freq=1, verbose=-1)

In [205]:
X_train[0]

array([-0.43383473, -0.15706168, -0.5777721 ,  0.39498994,  0.857798  ,
        0.22604842, -0.510776  ,  0.18485416,  0.7307111 ,  0.09243977,
        0.38286313,  0.8291649 ,  0.55900294, -0.8232045 , -0.1816736 ,
       -0.29083288, -0.05472206, -0.6253914 ,  0.7922106 , -0.15867963,
        0.88462347,  0.6400067 , -0.6405718 ,  0.08745352,  0.02386307,
        0.92898566, -0.35910046, -0.13943283,  0.3045259 ,  0.7808031 ,
       -0.6954054 ,  0.12978277,  0.15969662, -0.50586206,  0.40582216,
        0.87401867,  0.7884741 ,  0.14677425, -0.06325906,  0.04988863,
        0.524725  ,  0.7130524 ,  0.8716964 , -0.23032518,  0.09725419,
        0.8507136 ,  0.03445869, -0.48791707,  0.63076454,  0.5268865 ,
        0.4409193 ,  0.27802035,  0.33215016, -0.43133283,  0.90746784,
       -0.17597364, -0.33738223,  0.10844449, -0.8875365 ,  0.650042  ,
       -0.58702767,  0.2734547 ,  0.69263744, -0.0707315 ,  0.06625193,
        0.80624604, -0.31337297,  0.6293528 , -0.2500633 ,  0.18

In [207]:
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score


preds = model.predict_proba(X_test)
print(f"ROC-AUC target = {roc_auc_score(y_test, preds[:, 0])}")

ROC-AUC target = 0.5163018537581274


# TRX GPT Baseline

In [19]:
import pandas as pd
import numpy as np
from ptls.data_load.iterable_processing_dataset import IterableProcessingDataset
from ptls.data_load import IterableChain
from datetime import datetime
from ptls.data_load.datasets.parquet_dataset import ParquetDataset, ParquetFiles
from ptls.data_load.iterable_processing.feature_filter import FeatureFilter
from ptls.data_load.iterable_processing.to_torch_tensor import ToTorch
import torch
from functools import partial
from torch.utils.data import DataLoader
from ptls.data_load.padded_batch import PaddedBatch
from ptls.data_load.utils import collate_feature_dict
from tqdm import tqdm
import ptls
import torch.nn.functional as F

class QuantilfyAmount(IterableProcessingDataset):
    def __init__(self, col_amt='amount', quantilies=None):
        super().__init__()
        self.col_amt = col_amt
        if quantilies is None:
            self.quantilies = [0., 267.6, 1198.65, 3667.2, 8639.8, 18325.7, 36713.2, 68950.3, 143969.1, 421719.1]
        else: 
            self.quantilies = quantilies
    
    def __iter__(self):
        for rec in self._src:
            features = rec[0] if type(rec) is tuple else rec
            amount = features[self.col_amt]
            if not amount is torch.Tensor:
                amount = torch.Tensor(amount)
            am_quant = torch.zeros(len(amount), dtype=torch.int)
            for i, q in enumerate(self.quantilies):
                am_quant = torch.where(amount>q, i, am_quant)
            features[self.col_amt] = am_quant
            yield features

In [20]:
from ptls.data_load.datasets import ParquetDataset
from ptls.data_load.iterable_processing.time_proc import TimeProc
import numpy as np

trx_train = ParquetDataset(
    # shuffle_files=True,
    data_files=[
        './spark_data/trx_dataset_fold/fold=0',
        './spark_data/trx_dataset_fold/fold=1',
        # './spark_data/trx_dataset_fold/fold=2',
        './spark_data/trx_dataset_fold/fold=3',
        # './spark_data/trx_dataset_fold/fold=4'
    ],
    i_filters=[
        DeleteNan(col_name='mcc_code'),
        TypeProc(col_name='transaction_amt'),
        # ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
        #     drop_feature_names=[
        #         'user_id',
        #     ]
        # ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'mcc_code': 349,
                'currency_rk': 9
            }
        ),
        QuantilfyAmount(col_amt='transaction_amt'),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

# trx_valid = ParquetDataset(
#     shuffle_files=False,
#     data_files=[
#         './spark_data/trx_dataset_fold/fold=4'
#     ],
#     i_filters=[
#         DeleteNan(col_name='mcc_code'),
#         TypeProc(col_name='transaction_amt'),
#         ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
#             drop_feature_names=[
#                 'user_id',
#             ]
#         ),
#         ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
#         ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
#         TimeProc(time_col='event_time'),
#         ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
#             category_max_size={
#                 'mcc_code': 349,
#                 'currency_rk': 9
#             }
#         ),
#         QuantilfyAmount(col_amt='transaction_amt'),
#         ptls.data_load.iterable_processing.ToTorch()
#     ]
# )

In [21]:
from ptls.frames.coles import ColesIterableDataset
from ptls.frames.gpt.gpt_dataset import GptIterableDataset

trx_data_module = ptls.frames.PtlsDataModule(
    train_data=GptIterableDataset(
        data=trx_train,
        min_len=40,
        max_len=500
    ),
    train_batch_size=256,
    train_num_workers=0,
)

2025-05-15 05:11:31.179791: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1747285891.376946      35 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1747285891.430993      35 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [22]:
import pytorch_lightning as pl
import torch
from torch import nn
import warnings
from torchmetrics import MeanMetric
from typing import Tuple, Dict, List, Union

from ptls.nn.seq_encoder.abs_seq_encoder import AbsSeqEncoder
from ptls.nn import PBL2Norm
from ptls.data_load.padded_batch import PaddedBatch
from ptls.custom_layers import StatPooling, GEGLU
from ptls.nn.seq_step import LastStepEncoder

class Head(nn.Module):   
    def __init__(self, input_size, n_classes, hidden_size=64, drop_p=0.1):
        super().__init__()
        self.head = nn.Sequential(
            nn.Linear(input_size, hidden_size, bias=True),
            nn.GELU(),
            nn.Dropout(drop_p),
            nn.Linear(hidden_size, n_classes)
        )
    def forward(self, x):
        x = self.head(x)
        return x

class GptPretrainModule(pl.LightningModule):
    """GPT2 Language model

    Original sequence are encoded by `TrxEncoder`.
    Model `seq_encoder` predicts embedding of next transaction.
    Heads are used to predict each feature class of future transaction.

    Parameters
    ----------
    trx_encoder:
        Module for transform dict with feature sequences to sequence of transaction representations
    seq_encoder:
        Module for sequence processing. Generally this is transformer based encoder. Rnn is also possible
        Should works without sequence reduce
    head_hidden_size:
        Hidden size of heads for feature prediction
    seed_seq_len:
         Size of starting sequence without loss 
    total_steps:
        total_steps expected in OneCycle lr scheduler
    max_lr:
        max_lr of OneCycle lr scheduler
    weight_decay:
        weight_decay of Adam optimizer
    pct_start:
        % of total_steps when lr increase
    norm_predict:
        use l2 norm for transformer output or not
    inference_pooling_strategy:
        'out' - `seq_encoder` forward (`is_reduce_requence=True`) (B, H)
        'out_stat' - min, max, mean, std statistics pooled from `seq_encoder` layer (B, H) -> (B, 4H)
        'trx_stat' - min, max, mean, std statistics pooled from `trx_encoder` layer (B, H) -> (B, 4H)
        'trx_stat_out' - min, max, mean, std statistics pooled from `trx_encoder` layer + 'out' from `seq_encoder` (B, H) -> (B, 5H)
    """

    def __init__(self,
                 trx_encoder: torch.nn.Module,
                 seq_encoder: AbsSeqEncoder,
                 head_hidden_size: int = 64,
                 total_steps: int = 64000,
                 seed_seq_len: int = 16,
                 max_lr: float = 0.00005,
                 weight_decay: float = 0.0,
                 pct_start: float = 0.1,
                 norm_predict: bool = False,
                 inference_pooling_strategy: str = 'out_stat'
                 ):

        super().__init__()
        self.save_hyperparameters(ignore=['trx_encoder', 'seq_encoder'])

        self.trx_encoder = trx_encoder
        # assert self.trx_encoder.embeddings, '`embeddings` parameter for `trx_encoder` should contain at least 1 feature!'

        self._seq_encoder = seq_encoder
        self._seq_encoder.is_reduce_sequence = False

        self.head = nn.ModuleDict()
        for col_name, noisy_emb in self.trx_encoder.embeddings.items():
            self.head[col_name] = Head(input_size=self._seq_encoder.embedding_size, hidden_size=head_hidden_size, n_classes=noisy_emb.num_embeddings)

        if self.hparams.norm_predict:
            self.fn_norm_predict = PBL2Norm()

        self.loss = nn.CrossEntropyLoss(ignore_index=0)

        self.train_gpt_loss = MeanMetric()
        self.valid_gpt_loss = MeanMetric()

    def forward(self, batch: PaddedBatch):
        # print(f"{batch.payload['amount']}")
        z_trx = self.trx_encoder(batch) 
        out = self._seq_encoder(z_trx)
        if self.hparams.norm_predict:
            out = self.fn_norm_predict(out)
        return out
    
    def loss_gpt(self, predictions, labels, is_train_step):
        loss = 0
        for col_name, head in self.head.items():
            y_pred = head(predictions[:, self.hparams.seed_seq_len:-1, :])
            y_pred = y_pred.view(-1, y_pred.size(-1))

            y_true = labels[col_name][:, self.hparams.seed_seq_len+1:]
            y_true = torch.flatten(y_true.long())

            # print(f"{y_pred.size()=}")
            # print(f"{y_true=}")
            # print(y_true[10952])
            
            loss += self.loss(y_pred, y_true)
        return loss

    def training_step(self, batch, batch_idx):
        out = self.forward(batch)  # PB: B, T, H
        out = out.payload if isinstance(out, PaddedBatch) else out
        labels = batch.payload
        
        loss_gpt = self.loss_gpt(out, labels, is_train_step=True)
        self.train_gpt_loss(loss_gpt)
        self.log(f'gpt/loss', loss_gpt, sync_dist=True)
        return loss_gpt

    def validation_step(self, batch, batch_idx):
        out = self.forward(batch)  # PB: B, T, H
        out = out.payload if isinstance(out, PaddedBatch) else out
        labels = batch.payload

        loss_gpt = self.loss_gpt(out, labels, is_train_step=False)
        self.valid_gpt_loss(loss_gpt)

    def on_training_epoch_end(self):
        self.log(f'gpt/train_gpt_loss', self.train_gpt_loss, prog_bar=False, sync_dist=True, rank_zero_only=True)
        # self.train_gpt_loss reset not required here

    def on_validation_epoch_end(self):
        self.log(f'gpt/valid_gpt_loss', self.valid_gpt_loss, prog_bar=True, sync_dist=True, rank_zero_only=True)
        # self.valid_gpt_loss reset not required here

    def configure_optimizers(self):
        optim = torch.optim.NAdam(self.parameters(),
                                 lr=self.hparams.max_lr,
                                 weight_decay=self.hparams.weight_decay,
                                 )
        scheduler = torch.optim.lr_scheduler.OneCycleLR(
            optimizer=optim,
            max_lr=self.hparams.max_lr,
            total_steps=self.hparams.total_steps,
            pct_start=self.hparams.pct_start,
            anneal_strategy='cos',
            cycle_momentum=False,
            div_factor=25.0,
            final_div_factor=10000.0,
            three_phase=False,
        )
        scheduler = {'scheduler': scheduler, 'interval': 'step'}
        return [optim], [scheduler]
    
    @property
    def seq_encoder(self):
        return GPTInferenceModule(pretrained_model=self)

class GPTInferenceModule(torch.nn.Module):
    def __init__(self, pretrained_model):
        super().__init__()
        self.model = pretrained_model
        self.model.is_reduce_sequence = False

        self.stat_pooler = StatPooling()
        self.last_step = LastStepEncoder()

    def forward(self, batch):
        z_trx = self.model.trx_encoder(batch)
        out = self.model._seq_encoder(z_trx)
        out = out if isinstance(out, PaddedBatch) else PaddedBatch(out, batch.seq_lens)
        if self.model.hparams.inference_pooling_strategy=='trx_stat_out':
            stats = self.stat_pooler(z_trx)
            out = self.last_step(out)
            out = torch.cat([stats, out], dim=1)
        elif self.model.hparams.inference_pooling_strategy=='trx_stat':
            out = self.stat_pooler(z_trx)
        elif self.model.hparams.inference_pooling_strategy=='out_stat':
            out = self.stat_pooler(out)
        elif self.model.hparams.inference_pooling_strategy=='out':
            out = self.last_step(out)
        else:
            raise
        if self.model.hparams.norm_predict:
            out = out / (out.pow(2).sum(dim=-1, keepdim=True) + 1e-9).pow(0.5)
        return out

class CorrGptPretrainModule(GptPretrainModule):
    def __init__(self, feature_encoder, seq_encoder, trx_encoder, *args, **kwargs):
        trx_encoder.numeric_values = {}
        super().__init__(trx_encoder=trx_encoder, seq_encoder=seq_encoder, *args, **kwargs)
        self.save_hyperparameters(ignore=['trx_encoder', 'seq_encoder', 'feature_encoder'])
        self.feature_encoder = feature_encoder

    def forward(self, batch):
        z_trx = self.trx_encoder(batch)
        payload = z_trx.payload.view(z_trx.payload.shape[:-1] + (-1, 24))
        payload = self.feature_encoder(payload)
        encoded_trx = PaddedBatch(payload=payload, length=z_trx.seq_lens)
        out = self._seq_encoder(encoded_trx)
        if self.hparams.norm_predict:
            out = self.fn_norm_predict(out)
        return out

In [23]:
from ptls.frames.coles.metric import BatchRecallTopK
from functools import partial
import torch
from ptls.nn import TabFormerFeatureEncoder
from ptls.nn import TransformerEncoder
from ptls.nn import TrxEncoder
import ptls

feature_encoder = TabFormerFeatureEncoder(
    n_cols=5,
    emb_dim=24
)

seq_encoder = TransformerEncoder(
    n_heads=2,
    n_layers=2,
    input_size=120,
    use_positional_encoding=True
)

trx_encoder = ptls.nn.TrxEncoder(
    norm_embeddings=False,
    embeddings_noise=0.000,
    embeddings={
        'mcc_code': {'in': 350, 'out': 24},
        'currency_rk': {'in': 10, 'out': 24},
        'hour': {'in': 25, 'out': 24},
        'weekday': {'in': 8, 'out': 24},
        'transaction_amt': {'in': 11, 'out': 24}
    }
)

# seq_encoder = ptls.nn.RnnSeqEncoder(
#     trx_encoder=trx_encoder,
#     type='gru',
#     hidden_size=512
# )

pl_module = CorrGptPretrainModule(
    feature_encoder=feature_encoder
    trx_encoder=trx_encoder,
    seq_encoder=seq_encoder
)

In [24]:
!mkdir /kaggle/working/gpt_trx

In [25]:
from lightning.pytorch.loggers import WandbLogger

wandb_logger = WandbLogger(project="MBD_My_Code", log_model="all")

trainer = pl.Trainer(
    logger=wandb_logger,
    max_epochs=40,
    accelerator="cuda" if torch.cuda.is_available() else "cpu",
    enable_progress_bar=True,
    gradient_clip_val=0.5,
    log_every_n_steps=50,
    limit_val_batches=32,
)





INFO: Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
INFO: GPU available: True (cuda), used: True
INFO: TPU available: False, using: 0 TPU cores
INFO: HPU available: False, using: 0 HPUs


In [26]:
trainer.fit(pl_module, trx_data_module)

/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/configuration_validator.py:70: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.


/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.


Training: |          | 0/? [00:00<?, ?it/s]

INFO: `Trainer.fit` stopped: `max_epochs=40` reached.


In [28]:
trx_test = ParquetDataset(
    shuffle_files=False,
    data_files=[
        './spark_data/trx_dataset_fold/fold=4'
    ],
    i_filters=[
        DeleteNan(col_name='mcc_code'),
        TypeProc(col_name='transaction_amt'),
        # ptls.data_load.iterable_processing.feature_filter.FeatureFilter(
        #     drop_feature_names=[
        #         'user_id',
        #     ]
        # ),
        ptls.data_load.iterable_processing.SeqLenFilter(min_seq_len=32),
        ptls.data_load.iterable_processing.ISeqLenLimit(max_seq_len=4096),
        TimeProc(time_col='event_time'),
        ptls.data_load.iterable_processing.category_size_clip.CategorySizeClip(
            category_max_size={
                'mcc_code': 349,
                'currency_rk': 9
            }
        ),
        QuantilfyAmount(col_amt='transaction_amt'),
        ptls.data_load.iterable_processing.ToTorch()
    ]
)

In [31]:
from ptls.frames.inference_module import InferenceModule
from ptls.data_load.utils import collate_feature_dict

inference_module = InferenceModule(
        model=pl_module,
        pandas_output=True,
        drop_seq_features=True)

inference_dl = torch.utils.data.DataLoader(
    dataset=trx_test,
    collate_fn=collate_feature_dict,
    shuffle=False,
    batch_size=32,
    num_workers=2,

)

predict = trainer.predict(inference_module, inference_dl)

Predicting: |          | 0/? [00:00<?, ?it/s]

In [32]:
inference_embeddings = pd.concat(predict, axis=0)

In [39]:
train_matching = pd.read_csv('./data/raw_data/train_matching.csv')

def gen_random_for_negative(x, k, list_of_uniq):
    while True:
        final_list = np.random.choice(list_of_uniq, size=k, replace=False)
        if x not in final_list:
            return final_list
# ----
list_of_uniq_rtk = train_matching['rtk'].unique()
list_of_uniq_bank = train_matching['bank'].unique()
train_matching['negatives'] = train_matching['rtk'].apply(
                        lambda x: gen_random_for_negative(x, 3, list_of_uniq_rtk))
#-----

train_matching['target'] = 1
pos_part = train_matching[['bank', 'rtk', 'target']]
neg_part = train_matching[['bank', 'negatives']].explode('negatives')
neg_part['target'] = 0
neg_part.columns = pos_part.columns
full_dataset = pd.concat([pos_part, neg_part]).sort_values(by='bank')

from tqdm import tqdm

full_data = []
targets = full_dataset['target'].to_numpy()

for trx_id, click_id in tqdm(zip(full_dataset['bank'].to_numpy(), full_dataset['rtk'].to_numpy())):
    trx_emb = inference_embeddings[inference_embeddings['bank'] == str(trx_id)].drop(columns=['bank']).to_numpy().flatten()
    if len(trx_emb):
        full_data.append(trx_emb)

full_data = np.stack(full_data)

shuffled_ids = np.random.permutation(len(full_data))

train_ids = shuffled_ids[:int(len(shuffled_ids) * 0.8)]
test_ids = shuffled_ids[int(len(shuffled_ids) * 0.8) + 1]

X_train, y_train = full_data[train_ids], targets[train_ids]
X_test, y_test = full_data[test_ids], targets[test_ids]


0it [00:00, ?it/s]


KeyError: 'bank'

# Results

**Baseline**

ROC-AUC target = 0.57

**RnnEncoder with regional-attention**

reg = 5

ROC-AUC target = 0.59

**RnnEncoder with regional-attention**

reg = 10

ROC-AUC target = 0.53

**RnnEncoder with cross-attention**

sm = 5

big = 16

ROC-AUC target = 0.54

**RnnEncoder with cross-attention**

sm = 4

big = 15

ROC-AUC target = 0.54

**RnnEncoder with cross-attention**

sm = 3

big = 12

ROC-AUC target = 0.54

**Посмотреть распредлеение транзакций за неделю (взять несколько недель)**